Modules/grpmodule.c

```diff
@@ -119,16 +120,56 @@ grp_getgrgid_impl(PyObject *module, PyObject *id)
         }
         Py_DECREF(py_int_id);
     }
+#ifdef HAVE_GETGRGID_R
+    Py_BEGIN_ALLOW_THREADS
+    int status;
+    long bufsize = sysconf(_SC_GETGR_R_SIZE_MAX);
+    struct group grp;
+
+    buf = PyMem_RawMalloc(bufsize);
+
+    while (1) {
+        status = getgrgid_r(gid, &grp, buf, bufsize, &p);
+        if (p != NULL || status != ERANGE) {
+            break;
+        }
+        if (bufsize > PY_SSIZE_T_MAX) {
+            PyErr_NoMemory();
+            status = ERANGE;
+            break;
+        }
+        bufsize <<= 1;
+        p = PyMem_RawRealloc(buf, bufsize);
+        if (p == NULL) {
+            PyErr_NoMemory();
+            break;
+        }
+        buf = (char *) p;
+    }
 
-    if ((p = getgrgid(gid)) == NULL) {
+    if (status != 0) {
+        p = NULL;
```

Comment:
Is it possible that p != NULL if status != 0?

Reply:
Yes, p may be set to some arbitrary entry on some error conditions of getgrgid_r and similar methods.